src/riak_kv_ts_newtype.erl

```diff
@@ -239,13 +272,111 @@ add_ddl_ebin_to_path() ->
+            %% the table is up to date, no need to upgrade
+            ok;
+        true ->
+            %% upgrade, current version is greater than our persisted
+            %% known versions
+            {ok, DDL} = riak_kv_compile_tab:get_ddl(Table, HighestVersion),
+            DDLs = riak_ql_ddl:convert(CurrentVersion, DDL),
+            log_ddl_upgrade(DDL, DDLs),
+            [ok = riak_kv_compile_tab:insert(Table, DDLx) || DDLx <- DDLs],
+            DowngradedDDLs = riak_ql_ddl:convert(riak_ql_ddl:first_version(), DDL),
+            [ok = insert_downgraded_ddl(Table, DDLx) || DDLx <- DowngradedDDLs],
+            ok;
+        false ->
+            %% downgrade, the current version is lower than the latest
+            %% persisted version, we have to hope there is a downgraded
+            %% ddl in the compile tab
+            ok
+    end.
+
+insert_downgraded_ddl(BucketType, {error, {cannot_downgrade, Version}}) ->
+    log_cannot_downgrade_to_version(BucketType, Version);
+insert_downgraded_ddl(BucketType, DDL) ->
+    DDLVersion = riak_ql_ddl:ddl_record_version(element(1, DDL)),
+    case riak_kv_compile_tab:get_ddl(BucketType, DDLVersion) of
+        {ok,_} ->
+            %% the DDL already exists, so don't overwrite it
+            ok;
+        notfound ->
+            log_storing_downgraded_ddl(BucketType, DDLVersion),
+            riak_kv_compile_tab:insert(BucketType, DDL)
+    end.
+
 %%
--spec recompile_ddl(DDLVersion :: riak_ql_component:component_version()) -> ok.
-recompile_ddl(DDLVersion) ->
-    %% Get list of tables to recompile
-    Tables = riak_kv_compile_tab:get_ddl_records_needing_recompiling(DDLVersion),
-    lists:foreach(fun(Table) ->
-                      new_type(Table)
-                  end,
-                  Tables),
+log_storing_downgraded_ddl(BucketType, DDLVersion) when is_atom(DDLVersion) ->
+    lager:info("A DDL for table ~ts for DDL version ~p was stored, for use "
+               "in the event of a downgrade", [BucketType, DDLVersion]).
+
+%%
+log_cannot_downgrade_to_version(BucketType, Version) ->
+    lager:warning("Cannot downgrade table ~ts to version ~p, "
+                  "table will be disablded under this version",
+                  [BucketType, Version]).
+
+%%
+log_ddl_upgrade(DDL, DDLs) ->
+    lager:info("UPGRADING ~p WITH ~p", [DDL, DDLs]).
+
+%% For each table
+%%     build the table name
+%%     check if the beam is there for that module
+%%     if not then build it
+verify_helper_modules() ->
+    [verify_helper_module(T) || T <- riak_kv_compile_tab:get_all_table_names()],
     ok.
+
+verify_helper_module(Table) when is_binary(Table) ->
+    case beam_exists(Table) of
+        true ->
+            lager:info("beam file for table ~ts exists", [Table]),
+            ok;
+        false ->
+            lager:info("beam file for table ~ts must be recompiled", [Table]),
+            actually_compile_ddl(Table)
+    end.
+
+%%
+beam_exists(Table) when is_binary(Table) ->
```

Comment:
We would compile a new beam file and put it onto disk, but it wouldn't be loaded because the table helper modules are loaded naturally by the beam vm, the existing in-memory module would execute the function calls.

Since the module uses the compilers `vsn` attribute the compiler would almost certainly be exactly the same compiler that compiled the beam file that the nugget deleted.